{% extends "layout.html" %}
{% block head %}
	<!-- Load plotly.js into the DOM -->
	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>Activark (About)</title>
	<script src="{{url_for('static', filename='/js/ignoredFunctions.js')}}"></script>
	<script src='https://cdn.plot.ly/plotly-2.12.1.min.js'></script>
	<script src="https://code.jquery.com/jquery-3.6.0.js" integrity="sha256-H+K7U5CnXl1h5ywQfKtSj8PCmoN9aaq30gDh27Xc0jk=" crossorigin="anonymous"></script>
	<!-- DataTables -->
	<link rel="stylesheet" href="https://cdn.datatables.net/1.12.1/css/jquery.dataTables.min.css">
	<script src="https://cdn.datatables.net/1.12.1/js/jquery.dataTables.min.js"></script>
	<script src="https://cdn.datatables.net/buttons/2.3.6/js/dataTables.buttons.min.js"></script>
	<script src="https://cdnjs.cloudflare.com/ajax/libs/jszip/3.1.3/jszip.min.js"></script>
	<script src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.53/pdfmake.min.js"></script>
	<script src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.53/vfs_fonts.js"></script>
	<script src="https://cdn.datatables.net/buttons/2.3.6/js/buttons.html5.min.js"></script>
	<script src="https://cdn.datatables.net/buttons/2.3.6/js/buttons.print.min.js"></script>
	<!-- Bootstrap -->
	<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-0evHe/X+R7YkIZDRvuzKMRqM+OrBnVFBL6DOitfPri4tjfHxaWutUpFmBp4vmVor" crossorigin="anonymous">
	<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/js/bootstrap.bundle.min.js" integrity="sha384-pprn3073KE6tl6bjs2QrFaJGz5/SUsLqktiwsUTF55Jfv3qYSDhgCecCxMW52nD2" crossorigin="anonymous"></script>
{% endblock %}

{% block content %}
	<style>
			.center {
		display: block;
		margin-left: auto;
		margin-right: auto;
		width: 85%;
		}
	</style>
	<div class="container">
		<!-- Introduction -->
		<div id="introduction" class="row" style="padding-top: 50px; height: 500px;">
			<div class="card">
				<div class="card-body">
					<h3>Introduction</h3>
					<div class=".border-0" style="padding-top: 25px; width: 60%; outline: none !important; float: left;" data-step="2" data-title="Input information" data-intro="Information about the input">
						<p>
							Activark is a data-driven, ML-based approach to predict
							the functional consequence of genetic changes in protein kinases. 
							Activark was trained on a curated dataset 
							of activating (i.e. constitutive-activation or increase in kinase activity), 
							deactivating (i.e. loss or decrease in Kinase activity), and drug-resistance
							protein variants in human kinases and using sequence and structural features.
							The following sections describe the methodology used to develop Activark.
						</p>
						<p>
							To jump to a specific section on this page, click on the links below:
							<ul>
								<li><a href="#data_source">Data sources</a></li>
								<li><a href="#alignment">Constructing alignment and mapping known functional variants</a></li>
								<li><a href="#most_mutated">Quick observations</a></li>
								<li><a href="#feature_eng">Feature engineering and machine-learning</a></li>
								<li><a href="#Contacts">Contacts</a></li>
						</p>
					</div>
					<div class=".border-0" style="float: left;  height: 300px; width: 40%; outline: none !important; " data-step="2" data-title="Input information" data-intro="Information about the input">
						<span>
							<!-- <img src = "../static/img/source_plot.svg" height="400px" class="center" alt="workflow"/> -->
							<img src = "../static/img/logo_v1.svg" height="300px" class="center" alt="workflow"/>
						</span>
					</div>
					
				</div>
			</div>
		</div>
		<!-- Data source -->
		<div id="data_source" class="row" style="padding-top: 50px; height: 500px;">
			<div class="card">
				<div class="card-body">
					<h3>Data source</h3>
					<div class=".border-0" style="padding-top: 25px; width: 30%; outline: none !important; float: left;" data-step="2" data-title="Input information" data-intro="Information about the input">
						<span>
							We used <a href='https://www.uniprot.org/'>UniProt</a>,
							<a href='https://pubmed.ncbi.nlm.nih.gov/'>PubMed</a> and 
							<a href="https://cancer.sanger.ac.uk/cosmic">COSMIC</a> databases to obtain what we consider to be a 
							reliable set of variants that modulate kinase function. 
							We identified variants that led to:
							<ul>
								<li>constitutive-activation (activating)</li>
								<li>increase (activating)</li>
								<li>loss (deactivating)</li>
								<li>decrease (deactivating)</li> 
							</ul>
							in kinase activity, and those that led to 
							<ul>
								<li>drug resistance</li>
							</ul>
						</span>
					</div>
					<div class=".border-0" style="float: left;  height: 300px; width: 70%; outline: none !important; ">
						<span>
							<img src = "../static/img/source_plot.svg" height="350px" class="center" alt="workflow"/>
							<!-- <img src = "../static/img/logo_v1.svg" height="300px" class="center" alt="workflow"/> -->
						</span>
					</div>
					
				</div>
			</div>
		</div>
		<!-- Alignment/circular bar plot -->
		<div id="alignment" class="row" style="padding-top: 50px;">
			<div class="card">
				<div class="card-body">
					<h3>Constructing alignment and mapping known functional variants</h3>
					<span>
						We aligned the kinase sequences using the hmmalign tool from HMMER (version 3.1b2) 
						against their respective Pfam hidden Markov model profiles (Pkinase and PK_Tyr_Ser-Thr) 
						and trimmed the regions that were outside of the kinase domain. 
						We merged the two alignments using MAFFT (v7.520) and 
						mapped the known functional and neutral variants onto the alignment.
					</span>
					<span>
						<img src = "../static/img/main_figure.svg" width="100%" class="center" alt="workflow"/>
					</span>
				</div>
			</div>
		</div>
		<!-- Most mutations plot -->
		<div id="most_mutated" class="row" style="padding-top: 50px; height: 500px;">
			<div class="card">
				<div class="card-body">
					<h3>Quick observations</h3>
					<div class=".border-0" style="padding-top: 25px; width: 50%; outline: none !important; float: left;" >
						<span>
							Inspection of the variants revealed some obvious patterns.
							<ul>
								<li>Resistance and activating variants tend to avoid the most conserved and functionally core parts of the kinase catalytic domain and often overlap</li>
								<li>Loss-of-function or deactivating variants very often hit key parts of the enzymatic machinery, particularly the catalytic Lysine and the Aspartate residues in the “HRD'' and “DFG'' motifs</li>
								<li>Regions within the alignment where known activating and deactivating variants often overlap lie within the A-loop, and the N- or C-terminal tails of the kinase domain.</li>
							</ul>
						</span>
					</div>
					<div class=".border-0" style="float: left;  width: 50%; outline: none !important; ">
						<img src = "../static/img/most_mutated.svg" height="400px"; width="100%" class="center" alt="workflow"/>
					</div>
				</div>
			</div>
		</div>
		<!-- ML workflow -->
		<div id="feature_eng" class="row" style="padding-top: 50px;">
			<div class="card">
				<div class="card-body">
					<h3>Feature engineering and machine-learning</h3>
					<span>
						We applied a random forest algorithm to develop 3 contrasting predictors based on seven types of sequence and structural features:
						<ul>
							<li><kbd>Pred (A v D):</kbd> The first predictor, activating vs deactivating, 
								represents a typical situation when one has what is believed to be a 
								functional variant (e.g. observed many times in a cohort or dataset) 
								and wishes to distinguish these two possibilities.
							</li> 
							<li><kbd>Pred (A vs D vs N):</kbd> The second, activating, deactivating or neutral, is more reflective of a situation where 
								one does not know if a variant is functional at all and thus one needs to predict neutrals.
							</li> 
							<li><kbd>Pred (R vs N):</kbd>The third predictor, resistance vs neutral, predicts if a given mutation is resistant or not.</li>
						</ul>
						<strong>Important notes:</strong>
						<ul>
							<li>We avoided contrasting resistance to activating or deactivating due to the considerable overlap between activating 
								and resistance (above) and because resistance variants were from an entirely distinct source. </li>
							<li>This also means that predictions of resistance should probably be considered alongside the other two predictors, 
								since there will necessarily be a tendency to predict many activating or deactivating sites as resistant. </li> 
						</ul>

					</span>
					<img src = "../static/img/Figure_2_workflow_ML_only.svg" width="100%" class="center" alt="workflow"/>
				</div>
			</div>
		</div>

		<!-- Contacts -->
		<div id="Contacts" class="row" style="padding-top: 50px; padding-bottom: 100px;">
			<div class="card">
				<div class="card-body">
					<h3>Contacts</h3>
					<span>
						<kbd>Gurdeep Singh</kbd> gurdeep.singh[at]bioquant[dot]uni-heidelberg[dot]de <br>
						<kbd>Torsten Schmenger</kbd> torsten.schmenger[at]bioquant[dot]uni-heidelberg[dot]de <br>
						<kbd>Rob Russell</kbd> robert.russell[at]bioquant[dot]uni-heidelberg[dot]de <br>
					</span>
				</div>
			</div>
		</div>
		
		
	</div>
	  
{% endblock %}
