{% extends "layout.html" %}
{% block head %}
	<!-- Load plotly.js into the DOM -->
	<meta charset="utf-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>Activark (Help)</title>
	<script src="{{url_for('static', filename='/js/ignoredFunctions.js')}}"></script>
	<script src='https://cdn.plot.ly/plotly-2.12.1.min.js'></script>
	<script src="https://code.jquery.com/jquery-3.6.0.js" integrity="sha256-H+K7U5CnXl1h5ywQfKtSj8PCmoN9aaq30gDh27Xc0jk=" crossorigin="anonymous"></script>
	<!-- DataTables -->
	<link rel="stylesheet" href="https://cdn.datatables.net/1.12.1/css/jquery.dataTables.min.css">
	<script src="https://cdn.datatables.net/1.12.1/js/jquery.dataTables.min.js"></script>
	<script src="https://cdn.datatables.net/buttons/2.3.6/js/dataTables.buttons.min.js"></script>
	<script src="https://cdnjs.cloudflare.com/ajax/libs/jszip/3.1.3/jszip.min.js"></script>
	<script src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.53/pdfmake.min.js"></script>
	<script src="https://cdnjs.cloudflare.com/ajax/libs/pdfmake/0.1.53/vfs_fonts.js"></script>
	<script src="https://cdn.datatables.net/buttons/2.3.6/js/buttons.html5.min.js"></script>
	<script src="https://cdn.datatables.net/buttons/2.3.6/js/buttons.print.min.js"></script>
	<!-- Bootstrap -->
	<link href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-0evHe/X+R7YkIZDRvuzKMRqM+OrBnVFBL6DOitfPri4tjfHxaWutUpFmBp4vmVor" crossorigin="anonymous">
	<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.0-beta1/dist/js/bootstrap.bundle.min.js" integrity="sha384-pprn3073KE6tl6bjs2QrFaJGz5/SUsLqktiwsUTF55Jfv3qYSDhgCecCxMW52nD2" crossorigin="anonymous"></script>
{% endblock %}

{% block content %}
	<style>
			.center {
		display: block;
		margin-left: auto;
		margin-right: auto;
		width: 85%;
		}
	</style>
	<div class="container">
		<!-- Output -->
		<div id="output" class="row" style="padding-top: 50px;">
			<div class="card">
				<div class="card-body">
					<h3>Output page</h3>
					<div class=".border-0" style="padding-top: 25px; width: 60%; outline: none !important; float: left;">
						<span>
							<img src = "../static/img/outputTable.gif" class="center" alt="output"/>
						</span>
					</div>
					<div class=".border-0" style="float: left;   width: 40%; outline: none !important; ">
						<p>
							This page displays the brief output of the inputs:
							<ul>
							<li><kbd>Info:</kbd> Known functional and PTM information at the equivalent site in other kinases.</li>
							<li><kbd>View:</kbd> Click <i class="bi bi-box-arrow-in-up-right"></i> to view more descrition of the variant in another tab.</li>
							<li><kbd>Gene:</kbd> Gene name of the kinase.</li>
							<li><kbd>Variant:</kbd> Input variant.</li>
							<li><kbd>Kinase domain region:</kbd> Kinase domain region.</li>
							<li><kbd>Known functional consequence:</kbd> Any known functional consequence of the input variant.</li>
							<li><kbd>Verdict (activating vs deactivating):</kbd> 
								Likelihood if the given variant is activating or deactivating the kinase. 
								Values in paranthesis refer to the prediction confidence (Low, Medium, High). 
								Variants with uncertain verdicts are marked as "ambigous" and the ones that
								lie outside the kinase domain as '-'.</li>
							<li><kbd>Verdict (resistance):</kbd> 
								Likelihood if the given variant is resistant the kinase. 
								Values in paranthesis refer to the prediction confidence (Low, Medium. High). 
								Variants that lie outside the kinase domain as '-'.</li>
							</ul>
						</p>
					</div>
					
				</div>
			</div>
		</div>
		<!-- Summary of the variant -->
		<div id="summary" class="row" style="padding-top: 50px;">
			<div class="card">
				<div class="card-body">
					<h3>Summary of the variant</h3>
					<div class=".border-0" style="padding-top: 25px; width: 100%; outline: none !important; float: left;">
						<span>
							<img src = "../static/img/summaryChart.gif" height="400px" class="center" alt="summary"/>
						</span>
					</div>
					<!-- <div class=".border-0" style="float: left;   width: 40%; outline: none !important; "> -->
					<div class=".border-0" style="padding-top: 100px; outline: none !important; ">
						<p>
							This section displays the summary of the input variant, predicted probabilities and known functional
							and PTM information of the variant site in all kinases.
							On the left side of the section, brief summary of the input is displayed. On the right side, the following
							3 plots are displayed:
							<ul>
							<li><kbd>Prediction (Activating vs Deactivating)</kbd>: These predictions represent a typical situation 
								when one has what is believed to be a functional variant (e.g. observed many times in a cohort or dataset) 
								and wishes to distinguish these two possibilities.
								</li>
							<li><kbd>Prediction (Activating vs Deactivating vs Neutral)</kbd>: These predictions are more 
								reflective of a situation where one does not know if a variant is functional at all 
								and thus one needs to predict neutrals.</li>
							<li><kbd>Prediction (Resistance vs Not)</kbd>: This predicts if a given variant is resistant or not.
								<strong>Note</strong>: Given that this predictor lacks information about activating and deactivating information,
								users may ignore its predictions when other predictors predict the variant to be
								strongly deactivating (eg: variants occuring at most conserved residues like the Catalytic Lysine,
								DFG-motif, etc., which are associated with loss of kinase activity.)</li>
							</ul>
						</p>
						<p>
							<kbd>Known functional and PTM information in all kinases</kbd>: The table below (scroll) shows the known functional and PTM information 
								of the variant site and adjacent positions (+/- 2 in the alignment) in all kinases.</li>
						</p>
					</div>
					
				</div>
			</div>
		</div>
		<!-- Alignment -->
		<div id="alignment" class="row" style="padding-top: 50px;">
			<div class="card">
				<div class="card-body">
					<h3>Alignment</h3>
					<div class=".border-0" style="padding-top: 25px; width: 100%; outline: none !important; float: left;">
						<span>
							<img src = "../static/img/alignmentRun.gif" height="500px" class="center" alt="summary"/>
						</span>
					</div>
					<!-- <div class=".border-0" style="float: left;   width: 40%; outline: none !important; "> -->
					<div class=".border-0" style="padding-top: 100px; outline: none !important; ">
						<p>
							This section displays the alignment of the input kinase sequence with
							other human kinase sequences. Known functional variants are then mapped
							onto the alignment. Hover on the variants to know more about them.<br>
							<ul>
							<li><kbd>Window Size</kbd> represents the number of positions before or
								after the variant site to be displayed in the alignment (recommended 10-30).
								</li>
							<li><kbd>Show top N kinases</kbd> reflects the top N kinases to be displayed in the alignment.
								(recommended 10-100).
								</li>
							<li><kbd>Sort alignment by</kbd> represents the information (known functional or PTM)
								based on which the alignment will be sorted.
								</li>
							</ul>
						</p>
						You can view the alignment with all the kinases and covering the entire domain <a href="http://activark.russelllab.org/alignment"> here</a>.
					</div>
				</div>
			</div>
		</div>
		<!-- Contacts -->
		<div id="Contacts" class="row" style="padding-top: 50px; padding-bottom: 100px;">
			<div class="card">
				<div class="card-body">
					<h3>Contacts</h3>
					<span>
						<kbd>Gurdeep Singh</kbd> gurdeep.singh[at]bioquant[dot]uni-heidelberg[dot]de <br>
						<kbd>Torsten Schmenger</kbd> torsten.schmenger[at]bioquant[dot]uni-heidelberg[dot]de <br>
						<kbd>Rob Russell</kbd> robert.russell[at]bioquant[dot]uni-heidelberg[dot]de <br>
					</span>
				</div>
			</div>
		</div>
	</div>
	  
{% endblock %}
